### ISSUE TITLE
Unable to install with Ubuntu version of nim

### ISSUE THREAD

**Issue by schneiderfelipe** [inim-repl/INim#120]
Trying to install inim with nim 1.0.6 ([provided by Ubuntu 20.04.2 LTS](https://packages.ubuntu.com/focal/nim)), I get the following error: `Error: undeclared field: 'isEmptyOrWhitespace' for type system.string [declared in /usr/lib/nim/system.nim(49, 3)]`.

Maybe a minimum language version should be added to https://github.com/inim-repl/INim/blob/53a18a52f6a427ea273182ebb658ad2a28a7d333/inim.nimble#L12-L16

The offending line is https://github.com/inim-repl/INim/blob/53a18a52f6a427ea273182ebb658ad2a28a7d333/inim.nim#L207

## Version info, etc.

The following is what is provided by `apt` in Ubuntu:

```bash
$ nim --version
Nim Compiler Version 1.0.6 [Linux: amd64]
Compiled at 2020-02-27
Copyright (c) 2006-2019 by Andreas Rumpf

active boot switches: -d:release

$ nimble --version
nimble v0.11.0 compiled at 2020-02-27 15:07:29
git hash: couldn't determine git hash

```

## Full output

```bash
$ nimble install inim --verbose
    Reading official package list
Downloading https://github.com/inim-repl/INim using git
    Cloning latest tagged version: v0.6.1
  Verifying dependencies for inim@0.6.1
    Reading official package list
   Checking for cligen@>= 1.2.0
      Info: Dependency on cligen@>= 1.2.0 already satisfied
  Verifying dependencies for cligen@1.5.4
    Reading official package list
   Checking for noise@>= 0.1.4
      Info: Dependency on noise@>= 0.1.4 already satisfied
  Verifying dependencies for noise@0.1.4
 Installing inim@0.6.1
   Building inim/inim using c backend
    Prompt: Build failed for 'inim@0.6.1', would you like to try installing 'inim@#head' (latest unstable)? [y/N]
    Answer: y
    Reading official package list
Downloading https://github.com/inim-repl/INim using git
   Warning: Package 'inim' has an incorrect structure. It should contain a single directory hierarchy for source files, named 'inimpkg', but file 'test_piping_with_end_echo.nim' is in a directory named 'tests' instead. This will be an error in the future.
      Hint: If 'tests' contains source files for building 'inim', rename it to 'inimpkg'. Otherwise, prevent its installation by adding `skipDirs = @["tests"]` to the .nimble file.
  Verifying dependencies for inim@#head
    Reading official package list
   Checking for cligen@>= 1.2.0
      Info: Dependency on cligen@>= 1.2.0 already satisfied
  Verifying dependencies for cligen@1.5.4
    Reading official package list
   Checking for noise@>= 0.1.4
      Info: Dependency on noise@>= 0.1.4 already satisfied
  Verifying dependencies for noise@0.1.4
 Installing inim@#head
   Building inim/inim using c backend
     Error: Build failed for package: inim
        ... Details:
        ... Execution failed with exit code 1
        ... Command: "/usr/bin/nim" c --noNimblePath -d:NimblePkgVersion=0.6.1 -d:release --path:"/home/schneider/.nimble/pkgs/cligen-1.5.4"  --path:"/home/schneider/.nimble/pkgs/noise-0.1.4"  -o:"/tmp/nimble_18336/githubcom_inimreplINim_#head/inim" "/tmp/nimble_18336/githubcom_inimreplINim_#head/inim.nim"
        ... Output: Hint: used config file '/etc/nim/nim.cfg' [Conf]
        ... Hint: system [Processing]
        ... Hint: widestrs [Processing]
        ... Hint: io [Processing]
        ... Hint: inim [Processing]
        ... Hint: os [Processing]
        ... Hint: strutils [Processing]
        ... Hint: parseutils [Processing]
        ... Hint: math [Processing]
        ... Hint: bitops [Processing]
        ... Hint: macros [Processing]
        ... Hint: algorithm [Processing]
        ... Hint: unicode [Processing]
        ... Hint: pathnorm [Processing]
        ... Hint: osseps [Processing]
        ... Hint: posix [Processing]
        ... Hint: times [Processing]
        ... Hint: options [Processing]
        ... Hint: typetraits [Processing]
        ... Hint: osproc [Processing]
        ... Hint: strtabs [Processing]
        ... Hint: hashes [Processing]
        ... Hint: streams [Processing]
        ... Hint: cpuinfo [Processing]
        ... Hint: linux [Processing]
        ... Hint: strformat [Processing]
        ... Hint: terminal [Processing]
        ... Hint: colors [Processing]
        ... Hint: termios [Processing]
        ... Hint: sequtils [Processing]
        ... Hint: parsecfg [Processing]
        ... Hint: lexbase [Processing]
        ... Hint: tables [Processing]
        ... Hint: sugar [Processing]
        ... Hint: noise [Processing]
        ... Hint: basic [Processing]
        ... Hint: wtf8 [Processing]
        ... Hint: wcwidth [Processing]
        ... Hint: styler [Processing]
        ... Hint: posixio [Processing]
        ... Hint: prompt [Processing]
        ... Hint: history [Processing]
        ... Hint: killring [Processing]
        ... Hint: deques [Processing]
        ... /tmp/nimble_18336/githubcom_inimreplINim_#head/inim.nim(207, 52) Error: undeclared field: 'isEmptyOrWhitespace' for type system.string [declared in /usr/lib/nim/system.nim(49, 3)]
```

**Comment by schneiderfelipe** [comment]
Just confirming that it is a nim version issue, as with 1.4.8 inim installs without any issues:

```bash
$ nimble install inim --verbose
    Reading official package list
Downloading https://github.com/inim-repl/INim using git
    Cloning latest tagged version: v0.6.1
  Verifying dependencies for inim@0.6.1
    Reading official package list
   Checking for cligen@>= 1.2.0
      Info: Dependency on cligen@>= 1.2.0 already satisfied
  Verifying dependencies for cligen@1.5.4
    Reading official package list
   Checking for noise@>= 0.1.4
      Info: Dependency on noise@>= 0.1.4 already satisfied
  Verifying dependencies for noise@0.1.4
 Installing inim@0.6.1
   Building inim/inim using c backend
  Executing /home/schneider/.nimble/bin/nim c --colors:on --noNimblePath -d:release -d:NimblePkgVersion=0.6.1 --path:/home/schneider/.nimble/pkgs/cligen-1.5.4 --path:/home/schneider/.nimble/pkgs/noise-0.1.4 -o:/tmp/nimble_24190/githubcom_inimreplINim/inim /tmp/nimble_24190/githubcom_inimreplINim/src/inim.nim
Hint: used config file '/home/schneider/.choosenim/toolchains/nim-1.4.8/config/nim.cfg' [Conf]
Hint: used config file '/home/schneider/.choosenim/toolchains/nim-1.4.8/config/config.nims' [Conf]
............................................
/tmp/nimble_24190/githubcom_inimreplINim/src/inim.nim(420, 9) Hint: 'n' is declared but not used [XDeclaredButNotUsed]
/tmp/nimble_24190/githubcom_inimreplINim/src/inim.nim(457, 11) Hint: 'n' is declared but not used [XDeclaredButNotUsed]
/tmp/nimble_24190/githubcom_inimreplINim/src/inim.nim(543, 21) Hint: 'echo_status' is declared but not used [XDeclaredButNotUsed]
...........CC: stdlib_assertions.nim
CC: stdlib_io.nim
CC: stdlib_system.nim
CC: stdlib_parseutils.nim
CC: stdlib_math.nim
CC: stdlib_unicode.nim
CC: stdlib_strutils.nim
CC: stdlib_pathnorm.nim
CC: stdlib_posix.nim
CC: stdlib_times.nim
CC: stdlib_os.nim
CC: stdlib_hashes.nim
CC: stdlib_strtabs.nim
CC: stdlib_streams.nim
CC: stdlib_deques.nim
CC: stdlib_osproc.nim
CC: stdlib_strformat.nim
CC: stdlib_colors.nim
CC: stdlib_termios.nim
CC: stdlib_terminal.nim
CC: stdlib_lexbase.nim
CC: stdlib_tables.nim
CC: stdlib_parsecfg.nim
CC: ../../../../home/schneider/.nimble/pkgs/noise-0.1.4/noise/wtf8.nim
CC: ../../../../home/schneider/.nimble/pkgs/noise-0.1.4/noise/wcwidth.nim
CC: ../../../../home/schneider/.nimble/pkgs/noise-0.1.4/noise/basic.nim
CC: ../../../../home/schneider/.nimble/pkgs/noise-0.1.4/noise/posixio.nim
CC: ../../../../home/schneider/.nimble/pkgs/noise-0.1.4/noise/styler.nim
CC: ../../../../home/schneider/.nimble/pkgs/noise-0.1.4/noise/prompt.nim
CC: ../../../../home/schneider/.nimble/pkgs/noise-0.1.4/noise/history.nim
CC: ../../../../home/schneider/.nimble/pkgs/noise-0.1.4/noise/killring.nim
CC: ../../../../home/schneider/.nimble/pkgs/noise-0.1.4/noise.nim
CC: stdlib_critbits.nim
CC: ../../../../home/schneider/.nimble/pkgs/cligen-1.5.4/cligen/parseopt3.nim
CC: ../../../../home/schneider/.nimble/pkgs/cligen-1.5.4/cligen/mslice.nim
CC: ../../../../home/schneider/.nimble/pkgs/cligen-1.5.4/cligen/textUt.nim
CC: ../../../../home/schneider/.nimble/pkgs/cligen-1.5.4/cligen/argcvt.nim
CC: ../../../../home/schneider/.nimble/pkgs/cligen-1.5.4/cligen/sysUt.nim
CC: ../../../../home/schneider/.nimble/pkgs/cligen-1.5.4/cligen/humanUt.nim
CC: ../../../../home/schneider/.nimble/pkgs/cligen-1.5.4/cligen.nim
CC: inim.nim

Hint:  [Link]
Hint: 84233 lines; 4.316s; 145.363MiB peakmem; Release build; proj: /tmp/nimble_24190/githubcom_inimreplINim/src/inim.nim; out: /tmp/nimble_24190/githubcom_inimreplINim/inim [SuccessX]
    Copying file /tmp/nimble_24190/githubcom_inimreplINim/inim.nimble to /home/schneider/.nimble/pkgs/inim-0.6.1/inim.nimble
    Copying file /tmp/nimble_24190/githubcom_inimreplINim/inim to /home/schneider/.nimble/pkgs/inim-0.6.1/inim
   Creating symlink: /home/schneider/.nimble/pkgs/inim-0.6.1/inim -> /home/schneider/.nimble/bin/inim
   Success: inim installed successfully.

$ nim -v
Nim Compiler Version 1.4.8 [Linux: amd64]
Compiled at 2021-05-25
Copyright (c) 2006-2021 by Andreas Rumpf

git hash: 44e653a9314e1b8503f0fa4a8a34c3380b26fff3
active boot switches: -d:release

$ nimble -v
nimble v0.13.1 compiled at 2021-05-25 00:49:34
git hash: couldn't determine git hash
```

**Comment by joe733** [comment]
Yes, it installs properly.. but  I cannot run it:

```dart
~ 
➜ nimble install inim
Downloading https://github.com/inim-repl/INim using git
  Verifying dependencies for inim@0.6.1
      Info: Dependency on cligen@>= 1.2.0 already satisfied
  Verifying dependencies for cligen@1.5.4
      Info: Dependency on noise@>= 0.1.4 already satisfied
  Verifying dependencies for noise@0.1.4
 Installing inim@0.6.1
   Building inim/inim using c backend
    Prompt: inim@0.6.1 already exists. Overwrite? [y/N]
    Answer: y
   Success: inim installed successfully.

~ took 21s 
➜ inim 
zsh: command not found: inim

~ 
➜ nim -v                             
Nim Compiler Version 1.4.8 [Linux: amd64]
Compiled at 2021-06-07
Copyright (c) 2006-2021 by Andreas Rumpf

active boot switches: -d:release -d:nativeStackTrace

~ 
➜ nimble -v          
nimble v0.13.1 compiled at 2021-04-26 22:06:28
git hash: couldn't determine git hash
```

**Comment by schneiderfelipe** [comment]
@joe733 looks like a `PATH` issue in your case.

**Comment by joe733** [comment]
Yes, but I thought nimble would handle that. My system is Arch Linux, should I open a new issue mm?

This line https://github.com/inim-repl/INim/blob/53a18a52f6a427ea273182ebb658ad2a28a7d333/inim.nimble#L8 says it would install on such a location, but I do not know where that is.

My path includes: `... :/usr/local/bin:/usr/bin:/bin:/usr/local/sbin: ...` where I suppose `iNim` must be located. Cause..
```
```

**Comment by schneiderfelipe** [comment]
@joe733 I have `$HOME/.nimble/bin` in my `PATH`, and that's where `inim` is installed:

```bash
$ whence inim
/home/schneider/.nimble/bin/inim
```

Nim/Nimble installed with [`choosenim`](https://github.com/dom96/choosenim).

**Comment by joe733** [comment]
Okay, I've got that too.. I'll add it to the path. Thanks a lot!

**Comment by 0atman** [comment]
Thanks all, looks like we've learned a lot here today.

TLDR: inim (and perhaps the nim ecosystem in general) is only stable on modern versions of nim.

If you've got here from google, my recommendation is to uninstall your distro's version of nim, and install the latest with the official nim installer, choosenim. https://nim-lang.org/install_unix.html
